bindings/rust/s2n-tls/src/connection.rs

```diff
@@ -861,21 +868,28 @@ impl Connection {
         let handshake = unsafe {
             s2n_connection_get_handshake_type_name(self.connection.as_ptr()).into_result()?
         };
-        // The strings returned by s2n_connection_get_handshake_type_name
-        // are static and immutable after they are first calculated
-        static_const_str!(handshake)
+        // SAFETY: constructed strings have a null byte appended to them
+        // SAFETY: the data has a 'static lifetime, because it resides in a static
+        // char array, and is never modified after its initial creation
+        unsafe { const_str!(handshake) }
     }
 
     pub fn cipher_suite(&self) -> Result<&str, Error> {
         let cipher = unsafe { s2n_connection_get_cipher(self.connection.as_ptr()).into_result()? };
-        // The strings returned by s2n_connection_get_cipher
-        // are static and immutable since they are const fields on static const structs
-        static_const_str!(cipher)
+        // SAFETY: The data is null terminated because it is declared as a C string literal.
+        // SAFETY: cipher has a static lifetime because it lives on s2n_cipher_suite, a static struct.
+        // a static const struct, and therefore outlives the surrounding scope
+        // (which is the lifetime of the connection)
+        unsafe { const_str!(cipher) }
     }
 
     pub fn selected_curve(&self) -> Result<&str, Error> {
         let curve = unsafe { s2n_connection_get_curve(self.connection.as_ptr()).into_result()? };
-        static_const_str!(curve)
+        // SAFETY: The data is null terminated because it is declared as a C string literal.
+        // SAFETY: curve has a static lifetime because it lives on s2n_ecc_named_curve, which is a static const struct.
+        // a static const struct, and therefore outlives the surrounding scope
+        // (which is the lifetime of the connection)
+        unsafe { const_str!(curve) }
```

Comment:
> Don't these strings HAVE to be static, or they're unsafe to return?

No, the lifetime is inferred from the surrounding scope. Since the `&str` returned generally doesn't have a static lifetime, then there is no need for the underlying material to have a static lifetime. So in the context of `connection_something(&self)` this could be safely used on any material that lives for the lifetime of the connection.

My safety comments reference static lifetimes because the `'static` is obviously greater than the lifetime of `&self`, so it's an easy way to prove the second safety condition
> SAFETY: the underlying data must live at least as long as the surrounding scope